### PR TITLE
TLS 1.3: Resumption and early data compatibility tests

### LINKED ISSUE #6362: TLS 1.3 early data: ssl-opt.sh tests with GnuTLS and MbedTLS
The goal of this issue is to finalize the testing of the TLS 1.3 early data feature in ssl-opt.sh.

Current status:
Various tests with GnuTLS as a peer, client or server.
No test against OpenSSL but even in 3.x versions there does not seem to be support for it associated with re-connection using a ticket. As we support early data only with tickets, there is no way to test against OpenSSL in ssl-opt.sh without adding a new way of running tests in ssl-opt.sh. Not planned to do it yet, just GnuTLS as third party TLS 1.3 client/server for now.

Tasks:
- [ ] Clean-up of current early data tests against GnuTLS (check for redundancy and consistency, group them if this is not already the case).
- [ ] Add negative testing regarding setting of the peers: early data enabled in one peer but not in the other, early data enabled but the handshake go through sending an HRR.
- [ ] Add tests with MbedTLS as client and server 
- [ ] Add a negative test where a client try to send early data associated with a pre-shared key (not a ticket). We support eraly data only with tickets. Not sure if this is possible.
- [ ] Be sure https://github.com/Mbed-TLS/mbedtls/pull/8760#discussion_r1498901360 is addressed.

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Fix #6362

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required, will add one for TLS 1.3 early data feature as a whole
- [x] **backport** not required, no TLS 1.3 support in 2.28
- [x] **tests** provided

### CODE REVIEW

File: programs/ssl/ssl_server2.c [hunk 1]
```diff
@@ -1980,14 +1984,18 @@ int main(int argc, char *argv[])
...
 #if defined(MBEDTLS_SSL_EARLY_DATA)
-        else if (strcmp(p, "max_early_data_size") == 0) {
-            long long value = atoll(q);
-            tls13_early_data_enabled =
-                value >= 0 ? MBEDTLS_SSL_EARLY_DATA_ENABLED :
-                MBEDTLS_SSL_EARLY_DATA_DISABLED;
-            if (tls13_early_data_enabled) {
-                opt.max_early_data_size = atoi(q);
+        else if (strcmp(p, "early_data") == 0) {
+            switch (atoi(q)) {
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply programs/ssl/ssl_server2.c:1988]: The sample programs don't validate their options well. This is preexisting. The minor improvement in feedback for invalid command line options is not worth making a new option different from the others.

**Comment by ronald-cron-arm** [comment]
Following the merge of #8913 yesterday I had to rebase. The rebase was quite painful but the diff between the previous head https://github.com/Mbed-TLS/mbedtls/commit/ed3916ceacf1d2cc90bc76c3858da84ef11dfd59 and the new head https://github.com/Mbed-TLS/mbedtls/commit/dd2dc1578a2707d24ff516ca133def42f06bcaf4 for the 3 files changed in this PR: tls13-misc.sh, ssl-opt.sh and ssl_server2.c are clean thus this should be okay.

**Comment by tom-cosgrove-arm** [comment]
Is the CI failure something to worry about?

```
[2024-03-15T10:40:11.134Z] TLS 1.2: Check rsa_pss_rsae compatibility issue, m->O .................. tests/ssl-opt.sh: 1153: kill: No such process
[2024-03-15T10:40:11.134Z] 
[2024-03-15T10:40:11.134Z] RETRY(client-timeout) PASS
[2024-03-15T10:40:33.022Z] TLS 1.2: Check rsa_pss_rsae compatibility issue, m->G .................. tests/ssl-opt.sh: 1153: kill: No such process
[2024-03-15T10:40:33.022Z] 
[2024-03-15T10:40:33.022Z] RETRY(client-timeout) PASS
[2024-03-15T10:40:33.584Z] TLS 1.3 G->m: AES_128_GCM_SHA256,ffdhe3072,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:34.517Z] TLS 1.3 m->G: AES_128_GCM_SHA256,ffdhe3072,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:35.444Z] TLS 1.3 G->m: AES_128_GCM_SHA256,ffdhe4096,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:36.370Z] TLS 1.3 m->G: AES_128_GCM_SHA256,ffdhe4096,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:37.736Z] TLS 1.3 G->m: AES_128_GCM_SHA256,ffdhe6144,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:39.101Z] TLS 1.3 m->G: AES_128_GCM_SHA256,ffdhe6144,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:40.988Z] TLS 1.3 G->m: AES_128_GCM_SHA256,ffdhe8192,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:43.503Z] TLS 1.3 m->G: AES_128_GCM_SHA256,ffdhe8192,rsa_pss_rsae_sha256 ......... PASS
[2024-03-15T10:40:44.064Z] TLS 1.3: no HRR in case of PSK key exchange mode ....................... PASS
[2024-03-15T10:40:44.064Z] Handshake memory usage initial (MFL 16384 - default) ................... SKIP
[2024-03-15T10:40:44.064Z] Handshake memory usage (MFL 4096) ...................................... SKIP
[2024-03-15T10:40:44.064Z] Handshake memory usage (MFL 2048) ...................................... SKIP
[2024-03-15T10:40:44.064Z] Handshake memory usage (MFL 1024) ...................................... SKIP
[2024-03-15T10:40:44.064Z] Handshake memory usage (MFL 512) ....................................... SKIP
[2024-03-15T10:40:44.064Z] ------------------------------------------------------------------------
[2024-03-15T10:40:44.064Z] FAILED (1861 / 1862 tests (98 skipped))
[2024-03-15T10:40:44.320Z] ^^^^test_full_cmake_gcc_asan_new_bignum: test: ssl-opt.sh (full config, ASan build): tests/ssl-opt.sh -> 1^^^^
```

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM, as long as the CI failure is unrelated

**Review by waleed-elmelegy-arm — COMMENTED** [review]
Is it worth adding negative testing for max_early_data_size and ALPN with GnuTLS or is it out of scope? otherwise LGTM.

**Review by waleed-elmelegy-arm — APPROVED** [review]
LGTM

**Comment by ronald-cron-arm** [comment]
> LGTM, as long as the CI failure is unrelated

The test that failed is a DTLS test "Session resume using tickets, DTLS: openssl server", that has no relation with the changes in this PR. It passes on the open CI thus I do not think we have to worry about that.

**Comment by ronald-cron-arm** [comment]
> Is it worth adding negative testing for max_early_data_size and ALPN with GnuTLS or is it out of scope? otherwise LGTM.

That would involve setting configuration for the first handshake and then changing them when we resume. That's not really in the scope of ssl-opt.sh to me (and we have done that in test_suite_ssl).

**Pull request merged**